csa/man/acctdisk.8.in
.\"
.\" (C) 2004-2007 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This documentation supports a
.\" program that is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" The documentation that supports
.\" this program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this documentation and the software it
.\" supports, are distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\"
.\" Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
.\" Sunnyvale, CA  94085, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan
.\"
.TH acctdisk 8
.SH NAME
\f4acctdisk\f1 \- produces consolidated accounting records
.SH SYNOPSIS
.nf
\f4@sbindir@/acctdisk
.fi
.SH DESCRIPTION
\f4acctdisk\fP reads lines of information from standard input, converts
them to records in "Consolidated Accounting record" (\f3cacct\f1) format, and
writes out the records to standard output.
.PP
The input lines consist of the login ID and login name of the user and number of disk
blocks used by the user. These three sets of data are separated by
white space(s). Note that the login name information is not saved to cacct
records, but the input lines must contain that data.
.PP
The output is not in readable form. It is usually saved as a data
file for later use. These caact files can also be merged with \f4csaaddc\f1(8)
command.
.SH FILES
.nf
.ta 19
\f4/etc/passwd\fP	used for login name to user ID conversions
\f4@sbindir@/\fP	holds all administration accounting commands
\f4@csalocalstatedir@/day/pacct\fP	current process accounting file
.fi
.SH REFERENCES
chkconfig(8),
csaaddc(8),
dodisk(1M).


csa/ChangeLog
* Wed Jun  8 2016 Jay Lan <jay.j.lan@nasa.gov>
- Converted csa package from the last SGI-supported version 4.1.1 to
  community version 5.0.0

* Fri Oct 17 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.1
- Various fixes to allow for SLES11 build

* Fri Jul 25 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.0
- Fixed memory leaks
- Changed acctjob structure offset type from int to int64_t

* Fri Apr 18 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.2
- Updated to use new libnl API
- Fixed process exit code assignment

* Tue Mar 4 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.1
- Cleaned up acctdusg
- Added check to prevent SIGPIPE when communicating with daemon
- Increased maximum PID allowed for larger systems

* Wed Nov 7 2007 Jonathan Lim <jlim@sgi.com>
- csa-4.0.0
- Implement kernel functionality as a userspace daemon
- Remove obsolete wait metrics and replace with taskstats delay accounting
- Various bug fixes

* Fri Apr 27 2007 Jay Lan <jlan@sgi.com>
* Mon Apr 30 2007 Jonathan Lim <jlim@sgi.com>
- csa-3.0.2
- SBU's data was all zero in csa-3.0.0 and csa-3.0.1
- Clarify csacom(1) man page on '-f' option
- Obsolete csacom '-a' & '-q' option
- Fix a problem of csadrep reading input from a binary file
- Fixed csadrep(8) man page: replaced '-w' with '-n' and removed '-t'

* Wed Jul 5 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.1
- 'csacom -b' was broken in csa-3.0.0
- 'csacom -b' drops the last record of a pacct file

* Thu Feb 24 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.0
- csacom -[esES] option was broken.
- create /usr/sbin/rccsa symlink at %install in the spec file
- remove code for CSA record types not supported in Linux
- spliting up csa_kern.h into kernel version and userland version
- support pacct files created in csa-2.*.*

* Wed Jul 27 2005 Chris Sturtivant <csturtiv@sgi.com>
- csa-2.2.2
- Disabled perror for permission denied in csa_ctl when called by csa_auth

* Mon May 2 2005 Jay Lan <jlan@sgi.com>
- csa-2.2.1
- clean up csa init script
- csa_wracct(3) man page update
- 'csacom -o' wrote only one record
- 'csacom -b' does not work
- csajrep man page references invalid option '-W'
- 'csajrep -r' does not work
- 'dodisk' created files with incorrect user and group name 'adm'
- CSA does not handle gid's > 65535
- 'csacom -j' failed if job id greater than max value of an int

* Mon Nov 22 2004 Jay Lan <jlan@sgi.com>
- csa-2.2.0
- Fixed compilation issue with gcc 3.3.3
- CSA assumes existence of 'adm' user/group account
- csacom & csacms usage otuput missing carriage return
- csarecy misses a carriage return on certain condition
- csacom -W segfaults after printing one bad time

* Fri Oct 8 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.2
- Fixed 'csaswitch -c halt' bug
- Fixed 'csaedit dumped core on bad wkmg record' bug

* Tue Jul 27 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.1
- Remove static API library libcsa.a from rpm package
- Fixed a bug at csacom with invalid -u option
- cascom headers and data not aligned correctly at output

* Wed Jun 30 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.0
- Fixed a problem dealing with accounting files > 2GB
- Now derives "Day of Year" from other fields of the holidays file
- Removed "Pages Swapped" from csacom and csacms - not supported in Linux
- Fixed a bug of not mapping UID 0 to root

* Tue Feb 17 2004 Jay Lan <jlan@sgi.com>
- csa-2.0.0
- Implemented /proc ioctl interface
- Renamed rpm from csa-job-acct to csa
- Renamed /etc/init.d/csaacct to /etc/init.d/csa

* Fri Nov 7 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-2.0.0
- Provided API library libcsa.so and associated header file

* Wed Sep 3 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-1.0.3
- Updated the spec file

* Thu Feb 13 2003 Robin Holt <holt@sgi.com>
- Updated init scripts to use /var/lock/subsys

* Tue Nov 5 2002 Richard Offer <offer@sgi.com>
- Fixed packager/distribution/vendor

* Thu May 23 2002 Robin Holt <holt@sgi.com>
- csa-job-acct-1.0.2
- Changed install destination from /usr/local to /usr
- Updated the man page location to /usr/share/man
- Added a clean and realclean target to the Makefile

* Thu Apr 5 2001 Marlys Kohnke <kohnke@sgi.com>
- Changed csadrep internal markers back to match those in IRIX CSA
- Changed ja summary report so that all lines are displayed even if the
  resource used values are zero

* Sun Feb 11 2001 Marlys Kohnke <kohnke@sgi.com>
- Added support for Workload Management daemon accounting records
- Fixed csachkpacct script problem (df -k needs to be df -kP; changed awk
  field to 4)

* Fri Jan 19 2001 Daryl Grunau <dwg@acl.lanl.gov>
- port to Linux from SGI IRIX
- csa-job-acct-1.0.0 (available January 19, 2001)
- initial csa-job-acct-1.0.0 open source release


csa/cmd/acctdisk.c
/*
 * Copyright (c) 2004-2007 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston MA 02111-1307, USA.
 *
 * Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue
 * Sunnyvale, CA  94085, or:
 *
 * http://www.sgi.com
 */
#include <sys/types.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include "acctdef.h"
#include "acctmsg.h"
#include "cacct.h"

struct cacct cacct_record;

#define MAX_NAME_LEN 512
static char name[MAX_NAME_LEN]; /* unused at this time */
/* MAX_NAME_LEN - 1 to account for trailing NULL */
#define INPUT_FMT_STR "%ld %511s %lf"

/*
 * Read stdin, looking for uid, name, block tuples.
 */
int main(int argc, char *argv[])
{
	int input_status, cacct_fd;

	if ((cacct_fd = openacct(NULL, "w")) == -1) {
		acct_perr(ACCT_FATAL, errno, "Error occured opening stdout");
		exit(1);
	}

	memset((char *)&cacct_record, '\0', sizeof(struct cacct));

	if (create_hdr1(&cacct_record.ca_hdr, ACCT_CACCT) == -1) {
		acct_perr(ACCT_FATAL, errno, "Error occured creating record");
		exit(1);
	}

	cacct_record.ca_dc = 1;

	while((input_status = scanf(INPUT_FMT_STR, &cacct_record.ca_uid, name,
				    &cacct_record.ca_du)) != EOF) {
		if (input_status != 3) {
			fprintf(stderr, _("Error occurred scanning  '%s'."),
				"stdin");
			exit(1);
		}

		cacct_record.ca_prid = CACCT_NO_ID;
		cacct_record.ca_gid = CACCT_NO_ID;
		cacct_record.ca_jid = CACCT_NO_ID;
		cacct_record.ca_ash = CACCT_NO_ID;

		if (writecacct(cacct_fd, &cacct_record) !=
		    (sizeof(struct cacct))) {
			acct_perr(ACCT_FATAL, errno, "Error occurred writing "
				  "record");
			exit(1);
		}

		cacct_record.ca_uid = 0;
		cacct_record.ca_du = 0.0;
	}

	closeacct(cacct_fd);

	exit(0);
}
